src/org/sosy_lab/java_smt/solvers/bitwuzla/BitwuzlaModel.java

```diff
@@ -0,0 +1,238 @@
+    // Just declared (empty) arrays return an empty array
+
+    // Example Formula: (SELECT ARRAY INDEX)
+    // new ValueAssignment((SELECT ARRAY INDEX), value formula of INDEX, (SELECT ARRAY INDEX) =
+    // value of index, name of INDEX, actual java value of index, {return java value of the
+    // formula})
+
+    // TODO: Array equals value
+    Term array = pTerm;
+    Collection<ValueAssignment> assignments = new ArrayList<>();
+    Set<Term> indices = new HashSet<>();
+    while (array.sort().is_array()) {
+      // Here we either have STORE; SELECT or an empty Array
+      Vector_Term children = array.children();
+      List<Object> innerIndices = new ArrayList<>(upperIndices);
+      String name = getArrayName(array);
+      assert name != null;
+      if (children.isEmpty()) {
+        // Empty array
+        return assignments;
+      } else if (children.size() == 2) {
+        // SELECT expression
+        Term index = children.get(1);
+        if (!indices.add(index)) {
+          continue;
+        }
+
+        Term indexValue = bitwuzlaEnv.get_value(index);
+        innerIndices.add(evaluateImpl(indexValue));
+
+        // The value of an SELECT is equal to the content
+        Term valueTerm = bitwuzlaEnv.get_value(array);
+
+        assignments.add(
+            new ValueAssignment(
+                bitwuzlaCreator.encapsulateWithTypeOf(array),
+                bitwuzlaCreator.encapsulateWithTypeOf(valueTerm),
+                creator.encapsulateBoolean(buildEqForTwoTerms(array, valueTerm)),
+                name,
+                evaluateImpl(valueTerm),
+                innerIndices));
+        array = children.get(0);
+
+      } else {
+        // STORE expression
+        assert children.size() == 3;
+        Term index = children.get(1);
+        Term content = children.get(2);
+
+        if (!indices.add(index)) {
+          continue;
+        }
+
+        Term indexValue = bitwuzlaEnv.get_value(index);
+        Term contentValue = bitwuzlaEnv.get_value(content);
+        innerIndices.add(evaluateImpl(indexValue));
+        assignments.add(
+            new ValueAssignment(
+                bitwuzlaCreator.encapsulateWithTypeOf(array),
+                bitwuzlaCreator.encapsulateWithTypeOf(content),
+                creator.encapsulateBoolean(buildEqForTwoTerms(array, contentValue)),
+                name,
+                evaluateImpl(contentValue),
+                innerIndices));
+
+        array = children.get(0);
+        assignments.addAll(getArrayAssignments(array, innerIndices));
+      }
+    }
+    return assignments;
+  }
+
+  private String getArrayName(Term array) {
+    String name = array.symbol();
+    if (name != null) {
+      return name;
+    }
+    return getArrayName(array.get(0));
+  }
+
+  private Term buildEqForTwoTerms(Term left, Term right) {
+    assert left.sort().equals(right.sort());
+    Kind kind = Kind.EQUAL;
+    if (left.sort().is_fp()) {
+      kind = Kind.FP_EQUAL;
```

Comment:
just asking: for FP theory, the operation Kind.FP_Equal does handle NaN-equalities, right?

Reply:
It's poorly documented, but I think that Kind.FP_Equals sets -0=0 and nan/=nan, whereas it's -0/=0 and nan=nan for Kind.Equals.

Reply:
I've added a test.
The results are:
```
  // -0 == 0 for FP equals is SAT
  // -0 != 0 for equals
  // nan != nan for FP equals
  // nan == nan for equals is SAT
  // pos inf != neg inf FP eq
  // pos inf != neg inf eq
  ```